Adding the VAEAC approach

LHBO opened the pull request:
In this pull request, we add the `vaeac` (variational autoencoder with arbitrary conditioning) approach for estimating the conditional Shapley values. The `vaeac` approach allows for imputing missing values for arbitrary coalitions using a single variational autoencoder. Additionally, the approach supports mixed data. Finally, we have implemented the approach in [torch for R](https://torch.mlverse.org/).

The `vaeac` approach was introduced by [Olsen et al. (2022)](https://www.jmlr.org/papers/volume23/21-1413/21-1413.pdf), but they only provided a [python implementation](https://github.com/LHBO/ShapleyValuesVAEAC) built on top of the `shapr` package.

Key points in this pull request are the following:
1. Implemented the `vaeac` methodology in native `torch` in R.
2. Incorporated the `vaeac` approach into the `shapr` package as an additional method that supports mixed data.
3. Added `verbose` level to `shapr`, which allows the user to specify the degree of messages that will be given. This is in addition to progressr's progress bar.
4. Added functionality for sending a pre-trained `vaeac` model to `explain()` and for continuing to train the `vaeac` model for additional epochs if a user is unsatisfied with the accuracy. 
5. Added plot functions for monitoring the performance of the `vaeac` approach.
6. Added `plot_SV_several_approaches()`, which allows us to plot and compare several Shapley value explanations for the same explicands simultaneously.
7. Added a vignette specifically for the `vaeac` approach to demonstrate its capabilities and how to use it, in addition to some extra sections/paragraphs in the main vignette.
8. Changed how the vignettes are built to speed up the time. Now, we have to build them manually.
9. Added support for GPU. However, the CPU is often comparable/faster for smaller datasets.

martinju commented:
Ok, did not realize something is up with the OLD vignette, lines 831-851. Will look at this tomorrow.

martinju commented:
TODO to make checks pass:

- [x] Install torch
- [ ] Move from Boston MASS dataset to airquality (and rather make some variables categorical as in the tests). If we use the MASS package in the vignette, we need it as imports in DESCRIPTION, which we want avoid
- [ ] plot_vaeac_training_evaluation does not exists? For now I just removed it from the vignette for it to build. Needs to handle that properly.
- [ ] Add a vaeac.save_model option to explain which 
- [ ] Fix NSE warnings in checks

martinju commented on R/plot.R:
```diff
@@ -982,6 +982,10 @@ plot_MSEv_eval_crit <- function(explanation_list,
     stop("ggplot2 is not installed. Please run install.packages('ggplot2')")
   }
 
+  if (!requireNamespace("data.table", quietly = TRUE)) {
+    stop("data.table is not installed. Please run install.packages('data.table')")
+  }
+
```
I don't think this is necessary as data.table is already in imports of the shapley package, so not possible to install the shapr package without data.table.

LHBO replied:
Removed it here and in the MSEv plot function, too.

martinju commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+
+
+
+## Mixed Data
```
Let discuss whether we should include this or not. I did not check, but assume this by far the most time consuming part of the vignette since a bigger model is trained for longer. IF we decide to include it we should include the fitting time, and also have a very brief conclusion/discussion in the end where the performance, differences in the Shapley value and timing are compared. I guess the conclusion (for this specific use case) would then be that if you properly tune a big vaeac model you get a better-performing model than both the smaller vaeac model and the ctree model with default parameters. It should be mentioned that ctree also model parameters that can influence it's performance.

LHBO replied:
For me it does not matter if we include it or not. I have mainly used it to check that `vaeac` works and that is give approx same values as ctree.

martinju commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+
+
+# Future Updates
+-   Add support for GPU in vaeac. I have not had access to GPU, so have only used CPU.
```
I -> we.

LHBO replied:
Done

martinju reviewed (commented):
Comments for the vaeac vignette is provided below.

LHBO commented:
> TODO to make checks pass:
> 
> * [x]  Install torch
> * [x]  LARS: Move from Boston MASS dataset to airquality (and rather make some variables categorical as in the tests). If we use the MASS package in the vignette, we need it as imports in DESCRIPTION, which we want avoid
> * [x]  LARS: plot_vaeac_training_evaluation does not exists? For now I just removed it from the vignette for it to build. Needs to handle that properly.
> * [x]  LARS: Add a vaeac.save_model option to explain which
> * [x]  Fix NSE warnings in checks
> * [x]  LARS: Fix check warnings:
>   
>   * [x]  Missing link or links in documentation of:
>     
>     * [x]  SkipConnection: get_imputation_networks,
>     * [x]  Specified_masks_mask_generator: ‘row1, row1, row2, row2, row3, row3, ...’
>     * [x]  Specified_prob_mask_generator:     ‘row1, row1, row2, row2, row3, row3, ...’
>   * [x]  Undocumented arguments in documentation object 'vaeac_continue_train_model':     ‘explanation’ ‘...’
> * [x]  LARS: model.cuda in L1362 of approach.vaeac.R does not exists. What is the intention there?

Fixed all of these points. However, I am unsure if the package is installable here on Git. It runs fine on my Mac.

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+> [Future Updates](#FutureUpdates)
+
+<a id="intro"></a>
+
+# vaeac {#vaeac}
+
```
Done

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+---
+title: "The `vaeac` approach in `shapr`"
```
Fixed

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+y_train <- data[-ind_x_explain, get(y_var)]
+x_explain <- data[ind_x_explain, ..x_var]
+
+# Looking at the dependence between the features
+cor(x_train)
```
Removed

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+
+# Code Examples {#code}
+
+Here we go through how to use the `vaeac` approach.
```
DOne

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+
+
+
+## Vaeac with early stopping
```
Done

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
````diff
@@ -0,0 +1,981 @@
+```
+
+
+## Misplaced parameters
````
Removed it

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+vaeac.num_vaeacs_initiate <- 2
+
+# If we are to print the progress
+vaeac.verbose <- TRUE
```
Removed it

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+That is, each training batch contains two versions of the same training observation, but where the first one is
+masked by \eqn{S} and the second one is masked by \eqn{\bar{S}}, the complement, see
+\href{https://arxiv.org/pdf/2107.07436.pdf}{Jethani et al. (2022)}. 
+The training time will in due to more complex implementation in comparison to random sampling.
```
Rewritten

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+## Paired sampling
+
+The `vaeac` approach use what is called paired sampling to improve the stability of the training procedure.
+That is, each training batch contains two versions of the same training observation, but where the first one is
+masked by \eqn{S} and the second one is masked by \eqn{\bar{S}}, the complement, see
```
Fixed

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
````diff
@@ -0,0 +1,981 @@
+)
+```
+
+We can compare the results by looking at the training and validation errors and by the $\opertorname{MSE}_v$ evaluation criterion. 
````
Removed

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
````diff
@@ -0,0 +1,981 @@
+explanation_list <- list("Regular samp." = expl_paired_sampling_FALSE, "Paired samp." = expl_paired_sampling_TRUE)
+plot_several_vaeacs_VLB_IWAE(explanation_list)
+plot_MSEv_eval_crit(explanation_list)
+```
````
What do you mean? I don't follow. Arent the height of ALL plots set in knitr::opts_chunk$set? Is it this you want me to change?

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+
+
+
+## Progressr
```
I have kept it since we talked about verbose as a parameter to explain. I think this is something we can come back to in a another PR where we define what verbose = 1 and verbose = 2 actually means. Now I have set everything to verbose == 2, and I use message. Could use progressr with sticky, but then we need to be sure that a progress bar exists. So we should rather come back to this.

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
````diff
@@ -0,0 +1,981 @@
+
+
+```{r continue training, cache = TRUE}
+expl_litle_training <- explain(
````
Fixed thanks

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+
+## Continue the training of the vaeac approach
+
+In the case the user has set a too low number of training epochs and sees that the network is still learning,
+then the user can continue to train the network from where it stopped. Thus, a good workflow can therefore
+be to call the `explain()` function with a `n_samples = 1` (to not waste to much time to generate MC samples), 
+then look at the training and evaluation plots of the `vaeac`. If not satisfied, then train more. If satisfied,
```
Hmmm. Then explain will have a specific if clause just for vaeac. Would it be better if this was a global explain parameter? It would be nice to return the ctree models too. What do you think? I'll just keep it as it is. This is also one of the main reasons for why I wanted the train_vaeac function to be a separate function that is not too much connected to explain()/shapr.

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+
+
+
+## Grouping of features
```
I included this to see that the other masking scheme (i.e., not the MCAR) is used. I have made a point out of it now. We can remove it if wanted. I just included it see that vaeac works.

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
````diff
@@ -0,0 +1,981 @@
+  "Vaeac early stopping" = expl_early_stopping,
+  "Vaeac early stopping more epochs" = expl_early_stopping_train_more
+))
+```
+
````
DOne

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+plot_several_vaeacs_VLB_IWAE(list("Original" = expl_litle_training))
+
+# Can look at the saved models, training/validation errors, and the training time
+expl_litle_training$internal$parameters$vaeac$models
```
Removed

LHBO commented on vignettes/understanding_shapr_vaeac.Rmd:
```diff
@@ -0,0 +1,981 @@
+  explanation = expl_litle_training,
+  which_vaeac_model = "best",
+  true_data = x_train
+)
```
Moved

LHBO commented:
I fixed all comments sent by email and above. Refactored almost all the code and made separate functions for code that were identical in several places. We need to discuss verbose, but that should be done in a separate PR when we decide what the different levels actually mean.

Checked that all tests run (the check files must be updated) and added some global var in zzz file.

LHBO commented:
We must remember to update the text in the main vignette.  E.g. in "examples" section `vaeac` is not mentioned as a possible method.

LHBO commented on R/approach_vaeac.R:
```diff
@@ -131,414 +37,152 @@ setup_approach.vaeac <- function(internal, # add default values for vaeac here.
-  if (is.null(internal$parameters$vaeac.extra_parameters)) internal$parameters$vaeac.extra_parameters <- list()
+  # Check that torch is installed
+  if (!requireNamespace("torch", quietly = TRUE)) stop("`torch` is not installed. Please run install.packages('torch')")
+  if (!torch::torch_is_installed()) torch::install_torch()
```
Agree fixed

LHBO commented on R/approach_vaeac.R:
```diff
@@ -678,733 +245,423 @@ We set 'which_vaeac_model = best' and continue.\n",
 
-  # Include the number of trainable parameters in the state list.
-  state_list <- c(state_list, list("num_trainable_parameters" = model$num_train_param))
+  # Variables to stores the epochs of the `vaeac` at the best epoch according to IWAE and IWAE_running
```
Fixed

LHBO reviewed (commented):
Fixed all comments

martinju commented on vignettes/understanding_shapr.Rmd:
```diff
@@ -1270,7 +1270,7 @@ plot(explanation_combined, bar_plot_phi0 = FALSE, index_x_explain = c(1, 6))
 ![](figure_main/unnamed-chunk-20-1.png)
 
-As a second example using `"ctree"` to conditin on 1 and 2 features, and
+As a second example using `"ctree"` to conditing on 1 and 2 features, and
```
```suggestion
As a second example using `"ctree"` to condition on 1 and 2 features, and
```

martinju commented on vignettes/understanding_shapr.Rmd.orig:
````diff
@@ -1139,7 +1140,7 @@ explanation_combined <- explain(
 ```
 
-As a second example using `"ctree"` to conditin on 1 and 2 features, and
+As a second example using `"ctree"` to conditing on 1 and 2 features, and
````
```suggestion
As a second example using `"ctree"` to condition on 1 and 2 features, and
```

martinju reviewed (changes requested):
This looks good. Will just run a few test locally, before we merge. Seems that you fixed a typo with another typo. You can just click accept on these and the code will be updated directly on github.
